gradle-failure-reports/src/main/java/com/palantir/gradle/failurereports/CompileFailuresService.java

```diff
@@ -0,0 +1,162 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.failurereports;
+
+import com.google.common.base.Splitter;
+import com.palantir.gradle.failurereports.CompileFailuresService.Parameters;
+import com.palantir.gradle.failurereports.junit.JunitReporter;
+import com.palantir.gradle.failurereports.util.FailureReporterResources;
+import java.io.File;
+import java.nio.file.Path;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.gradle.api.Project;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+import org.gradle.api.tasks.compile.AbstractCompile;
+
+public abstract class CompileFailuresService implements BuildService<Parameters>, AutoCloseable {
+
+    interface Parameters extends BuildServiceParameters {
+        RegularFileProperty getOutputFile();
+
+        RegularFileProperty getCompileOutputFile();
+
+        Property<File> getRootDir();
+    }
+
+    private static final Pattern COMPILE_ERROR_FIRST_LINE_PATTERN =
+            Pattern.compile("^(?<sourcePath>.*):(?<lineNumber>\\d+): (?<errorMessage>error: .*)$");
+    private static final Pattern COMPILE_ERROR_LAST_LINE_PATTERN = Pattern.compile("^\\d+ error(s)*$");
+    private static final Pattern COMPILE_ERROR_PATTERN =
+            Pattern.compile("(?m)^(?<sourcePath>.*):(?<lineNumber>\\d+): (?<errorMessage>error: (.|\\s)*)");
+
+    private final ConcurrentMap<String, Boolean> startedCompileErrorsByTaskPath = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, StringBuilder> compilerErrorsByTaskPath = new ConcurrentHashMap<>();
+
+    public final <T extends AbstractCompile> void maybeCollectErrorMessage(T task, CharSequence charSequence) {
+        Matcher firstCompileErrorMatcher = COMPILE_ERROR_FIRST_LINE_PATTERN.matcher(charSequence);
+        if (firstCompileErrorMatcher.find()) {
+            String sourcePathFromError = firstCompileErrorMatcher.group("sourcePath");
+
+            // When running in parallel, javaCompileTasks will see the output of other tasks,
+            // see: https://github.com/gradle/gradle/issues/6068 for context.
+            // If the matched sourcePath is not part of the current javaCompileTask's sourceSet, we need to
+            // ignore it.
+            Set<String> taskCompiledSourcePaths = task.getSource().getFiles().stream()
+                    .map(File::getAbsolutePath)
+                    .collect(Collectors.toSet());
```

Comment:
Might make sense to lift this outside the loop? Otherwise we do this for every compile error.

Reply:
I tried to compute the sourcePaths for a compileJava task when I configure the stdOutputListener: https://github.com/palantir/gradle-failure-reports/pull/37/files/2e4c012c37db7c8dae00b432b7be341a07b1f13d#diff-7d99dff99bc546f36849ca5d92965dbafe8aa811b969528c58481e7f4244b350R48 but I am not getting all the paths. E.g. the testcase will only show one error instead of 2: https://github.com/palantir/gradle-failure-reports/pull/37/files#diff-37b2578a42112b856355fd9debe40e8c7090608890797a57a5850f1ef62bbb22R69. 

I used a "cachedSourcePaths" in the end: https://github.com/palantir/gradle-failure-reports/pull/37/commits/8f20791f905d7063a30bd3e12427dc3dfab8eadb#diff-7d99dff99bc546f36849ca5d92965dbafe8aa811b969528c58481e7f4244b350R53